layers.go

```diff
@@ -1200,16 +1197,19 @@ func (r *layerStore) PutAdditionalLayer(id string, parentLayer *Layer, names []s
 		r.byuncompressedsum[layer.UncompressedDigest] = append(r.byuncompressedsum[layer.UncompressedDigest], layer.ID)
 	}
 	if err := r.saveFor(layer); err != nil {
-		if err2 := r.driver.Remove(id); err2 != nil {
-			logrus.Errorf("While recovering from a failure to save layers, error deleting layer %#v: %v", id, err2)
+		if e := r.Delete(layer.ID); e != nil {
+			logrus.Errorf("While recovering from a failure to save layers, error deleting layer %#v: %v", id, e)
 		}
 		return nil, err
 	}
 	return copyLayer(layer), nil
 }
 
 // Requires startWriting.
-func (r *layerStore) Put(id string, parentLayer *Layer, names []string, mountLabel string, options map[string]string, moreOptions *LayerOptions, writeable bool, flags map[string]interface{}, diff io.Reader) (*Layer, int64, error) {
+func (r *layerStore) create(id string, parentLayer *Layer, names []string, mountLabel string, options map[string]string, moreOptions *LayerOptions, writeable bool, diff io.Reader) (layer *Layer, size int64, err error) {
```

Comment:
I completely forgot about that, but making `layer` a named return value apparently breaks the `defer` cleanup handler, because `return nil, -1, …` means “set `layer = nil`; then execute the `defer` handler”, and the handler will crash on trying to access `layer.ID`.

This was IIRC surprising to me, and a bit hard to find, in https://github.com/containers/storage/commit/ce3c7ffbd2e644f86d74cc7a51a35adb20fca6ed .

So I have developed quite an aversion to named return values…

Demo:
```go
package main

import "fmt"

type object struct{}

func unnamed() *object {
	i := &object{}
	defer func() { fmt.Printf("unnamed defer: %v\n", i) }()
	return nil
}

func named() (i *object) {
	i = &object{}
	defer func() { fmt.Printf("named defer: %v\n", i) }()
	return nil
}

func main() {
	fmt.Printf("unnamed return: %v\n", unnamed())
	fmt.Printf("named return: %v\n", named())
}
```

Reply:
Fortunately we have a local variable with the same value in it (we even use it in the error message on the next line, doh), so the change needed isn't a large one.